### ISSUE TITLE
Register custom browser protocols for .pdez and .pdex files

### ISSUE THREAD

**Issue by SableRaf** [benfry/processing4#559]
Since Processing 4.0, you can bundle sketches as double-clickable `.pdez` files and Libraries/Modes/Tools as `.pdex` files for easy sharing and installation. (see [Processing #73](https://github.com/processing/processing/issues/73), [Processing #3987](https://github.com/processing/processing/issues/3987), and the [release notes for 4.0 beta1](https://github.com/processing/processing4/blob/58ff6a3ca5c9c2f26a81d4014edc5d5ba0abf4c7/build/shared/revisions.md#processing-40-beta-1))

A great next step would be to allow `.pdez` and `.pdex` files to be opened in the PDE directly from a url in a web browser. This would enable easier sharing of sketches on the forum, better UX for multi-tab examples on the website, and open the possibility for online Processing sketch galleries.

This would require the user to register the protocol on their system (which I imagine could be done from the PDE ?) For example, [this page](https://developer.shotgridsoftware.com/af0c94ce/) describes the steps needed to register a custom browser protocol on Window, macOS and Linux. There is also a [W3C documentation page](https://www.w3.org/TR/2011/WD-html5-20110525/timers.html#custom-handlers) about custom scheme and content handlers.

**Comment by SableRaf** [comment]
On a side note, we may need to warn people to only run code when they trust the author (similar to [VSCode's restricted mode](https://code.visualstudio.com/docs/editor/workspace-trust))

![image](https://user-images.githubusercontent.com/290261/191957342-bf3b3556-ac38-4cf4-a8b6-36a07cfee03d.png)

**Comment by benfry** [comment]
That should be in place because the file extension mapping should be happening. Is that not working? 

For the second point, can you open a separate issue? Or perhaps even before that, have you tried it to see what it says now? I realize it's not making a stink about the security implications, but it should already be prompting whether you meant to open it or not.

**Comment by SableRaf** [comment]
It works as intended if you download the file then open it. 

I was talking about opening the file directly by clicking a link in the form `processing4://my-website.com/sketch220924.pdez` in the browser. This requires an extra step of registering the URI scheme at the OS level. If Processing already does that I couldn't find it in core.

**Comment by SableRaf** [comment]
I did some more research.

On macOS at least, the first step is adding the protocol name and scheme to the `info.plist` file. For example:

```html
<key>CFBundleURLTypes</key>
<array>
	<dict>
		<key>CFBundleURLName</key>
		<string>Processing 4</string>
		<key>CFBundleURLSchemes</key>
		<array>
			<string>processing4</string>
		</array>
	</dict>
</array>
```

Then there needs to be a handler for these URLs in the PDE code using `setOpenURIHandler` from what I understand (which is admittedly very little).

References:
- [Handle custom URL schemes in an OS X Java application](https://stackoverflow.com/questions/20074610/handle-custom-url-schemes-in-an-os-x-java-application)
- [Access parameters from custom URI scheme in Java application on OS X](https://stackoverflow.com/questions/27435606/access-parameters-from-custom-uri-scheme-in-java-application-on-os-x)

**Comment by benfry** [comment]
You can just use http:// or https:// URLs, no need to do a special protocol handler.

**Comment by SableRaf** [comment]
I made a quick test to show the current behavior: https://sableraf.github.io/testURIscheme/

`http://` or `https://` links trigger the browser to download the file (tested on Windows 10 and macOS 12.5.1)

Note: I also included a placeholder link with a `processingOrg://` custom protocol for future testing.

**Comment by benfry** [comment]
Implemented for macOS and Windows for 4.2; uses `pde://` and can point to a local file or an https url online. More to come…

**Comment by SableRaf** [comment]
Thanks @benfry this is amazing! Shoutout to @opyh for initially suggesting the idea in a conversation we had back in August.
